server/src/main/java/org/elasticsearch/search/aggregations/pipeline/PercentileRanksBucketPipelineAggregationBuilder.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.pipeline;
+
+import com.carrotsearch.hppc.DoubleArrayList;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.search.aggregations.AggregationBuilder;
+import org.elasticsearch.search.aggregations.AggregatorFactory;
+import org.elasticsearch.search.aggregations.PipelineAggregationBuilder;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+
+public class PercentileRanksBucketPipelineAggregationBuilder
+    extends BucketMetricsPipelineAggregationBuilder<PercentileRanksBucketPipelineAggregationBuilder> {
+    public static final String NAME = "percentile_ranks_bucket";
+    static final ParseField PERCENTS_FIELD = new ParseField("percents");
+    static final ParseField KEYED_FIELD = new ParseField("keyed");
+
+    private double[] percents = new double[] { 1.0, 5.0, 25.0, 50.0, 75.0, 95.0, 99.0 };
```

Comment:
No need for a default here.  Unlike a percentiles agg (which has to be 0 to 100), the user supplies the values for a percentile_ranks agg and we don't know what those values are.  They could be 0-1, or -100-99292.  Who knows :)

The percentile_ranks agg doesn't have a default, and makes the `values` field mandatory.

Reply:
Got it. 

So now, How shall`values` field be made manadatory?  have `values` as argument to `PercentileRanksBucketPipelineAggregationBuilder` class's constructor?

Reply:
I have one more question @polyfractal  on how `values` are retrieved into this aggregation system. 
Is it being peculated to here via Elasticsearch database when user fills in data entries and all user does is call this API to get `PercentileRanks` bucket 
or
values array/list is provided specifically while calling this specific API to retrieve bucket of percentileRanks ?

Reply:
would you also please point me to high level design document for elasticsearch codebase to help me better prepare for further contribution?